typescript/sdk/src/router/MultiProtocolRouterApps.ts

```diff
@@ -0,0 +1,85 @@
+import {
+  Address,
+  Domain,
+  ProtocolType,
+  promiseObjAll,
+} from '@hyperlane-xyz/utils';
+
+import { AdapterClassType, MultiProtocolApp } from '../app/MultiProtocolApp';
+import { ChainMap, ChainName } from '../types';
+
+import {
+  EvmGasRouterAdapter,
+  EvmRouterAdapter,
+} from './adapters/EvmRouterAdapter';
+import {
+  SealevelGasRouterAdapter,
+  SealevelRouterAdapter,
+} from './adapters/SealevelRouterAdapter';
+import { IGasRouterAdapter, IRouterAdapter } from './adapters/types';
+import { RouterAddress } from './types';
+
+export { Router } from '@hyperlane-xyz/core';
+
+export class MultiProtocolRouterApp<
+  ContractAddrs extends RouterAddress = RouterAddress,
+  IAdapterApi extends IRouterAdapter = IRouterAdapter,
+> extends MultiProtocolApp<ContractAddrs, IAdapterApi> {
+  // Explicit cast via unknown is unfortunately required here to enable both default adapters AND generic types
+  public readonly adapters = {
+    [ProtocolType.Ethereum]: EvmRouterAdapter as unknown as AdapterClassType<
+      ContractAddrs,
+      IAdapterApi
+    >,
+    [ProtocolType.Sealevel]:
+      SealevelRouterAdapter as unknown as AdapterClassType<
+        ContractAddrs,
+        IAdapterApi
+      >,
+  };
+
+  router(chain: ChainName): Address {
+    return this.metadata(chain).router;
+  }
+
+  interchainSecurityModules(): Promise<ChainMap<Address>> {
```

Comment:
kinda seems like we have a pretty consistent pattern?
is it possible to just like inherit the interface of `adapter[chain]`

Reply:
Per discord chat, I did try to make that approach work but it doesn't seem feasible due to typescript limitations